File: apps/backend/src/db/story_store_test.go
package db

import (
	"errors"
	"testing"

	"github.com/lucas/oraculo/apps/backend/src/domain"
)

func TestStoryStore_ListSummaries(t *testing.T) {
	database := testDB(t)
	epicStore := NewEpicStore(database)
	storyStore := NewStoryStore(database)
	taskStore := NewTaskStore(database)

	epic, _, err := epicStore.Create("test-epic", "desc")
	if err != nil {
		t.Fatalf("create epic: %v", err)
	}

	story, _, err := storyStore.Create(epic.ID, "story-1", "desc")
	if err != nil {
		t.Fatalf("create story: %v", err)
	}

	// Create 3 tasks: 2 completed, 1 failed
	for _, name := range []string{"task-1", "task-2", "task-3"} {
		if _, _, err := taskStore.Create(story.ID, name, "d", nil); err != nil {
			t.Fatalf("create %s: %v", name, err)
		}
	}

	if _, err := taskStore.Start(story.ID, "task-1"); err != nil {
		t.Fatalf("start task-1: %v", err)
	}
	if _, err := taskStore.Complete(story.ID, "task-1", domain.TaskResult{Summary: "done"}); err != nil {
		t.Fatalf("complete task-1: %v", err)
	}
	if _, err := taskStore.Start(story.ID, "task-2"); err != nil {
		t.Fatalf("start task-2: %v", err)
	}
	if _, err := taskStore.Complete(story.ID, "task-2", domain.TaskResult{Summary: "done"}); err != nil {
		t.Fatalf("complete task-2: %v", err)
	}
	if _, err := taskStore.Start(story.ID, "task-3"); err != nil {
		t.Fatalf("start task-3: %v", err)
	}
	if _, err := taskStore.Fail(story.ID, "task-3", "oops"); err != nil {
		t.Fatalf("fail task-3: %v", err)
	}

	summaries, err := storyStore.ListSummaries(epic.ID)
	if err != nil {
		t.Fatalf("ListSummaries: %v", err)
	}
	if len(summaries) != 1 {
		t.Fatalf("expected 1 summary, got %d", len(summaries))
	}
	s := summaries[0]
	if s.TaskCount != 3 {
		t.Errorf("TaskCount = %d, want 3", s.TaskCount)
	}
	if s.CompletedTaskCount != 2 {
		t.Errorf("CompletedTaskCount = %d, want 2", s.CompletedTaskCount)
	}
	if s.FailedTaskCount != 1 {
		t.Errorf("FailedTaskCount = %d, want 1", s.FailedTaskCount)
	}
}

// createEpic is a test helper that creates an epic and returns its ID.
func createEpic(t *testing.T, database *DB, name string) int {
	t.Helper()
	epic, _, err := NewEpicStore(database).Create(name, "test epic")
	if err != nil {
		t.Fatalf("createEpic(%q): %v", name, err)
	}
	return epic.ID
}

func TestStoryStore_Create(t *testing.T) {
	database := testDB(t)
	epicID := createEpic(t, database, "my-epic")
	store := NewStoryStore(database)

	story, created, err := store.Create(epicID, "my-story", "a description")
	if err != nil {
		t.Fatalf("Create: %v", err)
	}
	if !created {
		t.Error("expected created=true on first call")
	}
	if story.Name != "my-story" {
		t.Errorf("Name = %q, want %q", story.Name, "my-story")
	}
	if story.Description != "a description" {
		t.Errorf("Description = %q, want %q", story.Description, "a description")
	}
	if story.EpicID != epicID {
		t.Errorf("EpicID = %d, want %d", story.EpicID, epicID)
	}
}

func TestStoryStore_CreateIdempotent(t *testing.T) {
	database := testDB(t)
	epicID := createEpic(t, database, "my-epic")
	store := NewStoryStore(database)

	s1, _, _ := store.Create(epicID, "my-story", "desc")
	s2, created, err := store.Create(epicID, "my-story", "desc")
	if err != nil {
		t.Fatalf("second Create: %v", err)
	}
	if created {
		t.Error("expected created=false on second call")
	}
	if s2.ID != s1.ID {
		t.Errorf("IDs differ: %d vs %d", s2.ID, s1.ID)
	}
}

func TestStoryStore_CreateSameNameDifferentEpic(t *testing.T) {
	database := testDB(t)
	epicID1 := createEpic(t, database, "epic-one")
	epicID2 := createEpic(t, database, "epic-two")
	store := NewStoryStore(database)

	s1, created1, err := store.Create(epicID1, "shared-name", "desc1")
	if err != nil {
		t.Fatalf("Create under epic1: %v", err)
	}
	if !created1 {
		t.Error("expected created=true for epic1")
	}

	s2, created2, err := store.Create(epicID2, "shared-name", "desc2")
	if err != nil {
		t.Fatalf("Create under epic2: %v", err)
	}
	if !created2 {
		t.Error("expected created=true for epic2")
	}

	if s1.ID == s2.ID {
		t.Errorf("expected different IDs, both got %d", s1.ID)
	}
}

func TestStoryStore_GetByName(t *testing.T) {
	database := testDB(t)
	epicID := createEpic(t, database, "my-epic")
	store := NewStoryStore(database)

	store.Create(epicID, "test-story", "desc")
	story, err := store.GetByName(epicID, "test-story")
	if err != nil {
		t.Fatalf("GetByName: %v", err)
	}
	if story.Name != "test-story" {
		t.Errorf("Name = %q, want %q", story.Name, "test-story")
	}
	if story.EpicID != epicID {
		t.Errorf("EpicID = %d, want %d", story.EpicID, epicID)
	}
}

func TestStoryStore_GetByName_WrongEpic(t *testing.T) {
	database := testDB(t)
	epicID1 := createEpic(t, database, "epic-one")
	epicID2 := createEpic(t, database, "epic-two")
	store := NewStoryStore(database)

	store.Create(epicID1, "only-in-epic1", "desc")
	_, err := store.GetByName(epicID2, "only-in-epic1")
	if err == nil {
		t.Fatal("expected error when querying wrong epic")
	}
	if !errors.Is(err, domain.ErrNotFound) {
		t.Errorf("expected ErrNotFound, got: %v", err)
	}
}

func TestStoryStore_GetByName_NotFound(t *testing.T) {
	database := testDB(t)
	epicID := createEpic(t, database, "my-epic")
	store := NewStoryStore(database)

	_, err := store.GetByName(epicID, "nonexistent")
	if err == nil {
		t.Fatal("expected error for nonexistent story")
	}
	if !errors.Is(err, domain.ErrNotFound) {
		t.Errorf("expected ErrNotFound, got: %v", err)
	}
}

func TestStoryStore_List(t *testing.T) {
	database := testDB(t)
	epicID1 := createEpic(t, database, "epic-one")
	epicID2 := createEpic(t, database, "epic-two")
	store := NewStoryStore(database)

	store.Create(epicID1, "alpha", "")
	store.Create(epicID1, "beta", "")
	store.Create(epicID2, "gamma", "")

	stories, err := store.List(epicID1)
	if err != nil {
		t.Fatalf("List: %v", err)
	}
	if len(stories) != 2 {
		t.Fatalf("len = %d, want 2", len(stories))
	}
	if stories[0].Name != "alpha" {
		t.Errorf("stories[0].Name = %q, want %q", stories[0].Name, "alpha")
	}
	if stories[1].Name != "beta" {
		t.Errorf("stories[1].Name = %q, want %q", stories[1].Name, "beta")
	}
}

func TestStoryStore_Update(t *testing.T) {
	database := testDB(t)
	epicID := createEpic(t, database, "my-epic")
	store := NewStoryStore(database)

	store.Create(epicID, "my-story", "old")
	story, err := store.Update(epicID, "my-story", "new description")
	if err != nil {
		t.Fatalf("Update: %v", err)
	}
	if story.Description != "new description" {
		t.Errorf("Description = %q, want %q", story.Description, "new description")
	}
}

func TestStoryStore_Update_NotFound(t *testing.T) {
	database := testDB(t)
	epicID := createEpic(t, database, "my-epic")
	store := NewStoryStore(database)

	_, err := store.Update(epicID, "nonexistent", "desc")
	if !errors.Is(err, domain.ErrNotFound) {
		t.Errorf("expected ErrNotFound, got: %v", err)
	}
}

func TestStoryStore_Delete(t *testing.T) {
	database := testDB(t)
	epicID := createEpic(t, database, "my-epic")
	store := NewStoryStore(database)

	store.Create(epicID, "doomed", "")
	if err := store.Delete(epicID, "doomed"); err != nil {
		t.Fatalf("Delete: %v", err)
	}
	_, err := store.GetByName(epicID, "doomed")
	if !errors.Is(err, domain.ErrNotFound) {
		t.Error("expected ErrNotFound after delete")
	}
}

func TestStoryStore_Delete_NotFound(t *testing.T) {
	database := testDB(t)
	epicID := createEpic(t, database, "my-epic")
	store := NewStoryStore(database)

	err := store.Delete(epicID, "nonexistent")
	if !errors.Is(err, domain.ErrNotFound) {
		t.Errorf("expected ErrNotFound, got: %v", err)
	}
}

func TestStoryStore_UpdateApprovalStatus(t *testing.T) {
	database := testDB(t)
	epicID := createEpic(t, database, "my-epic")
	store := NewStoryStore(database)

	store.Create(epicID, "my-story", "")
	if err := store.UpdateApprovalStatus(epicID, "my-story", domain.ApprovalPending); err != nil {
		t.Fatalf("UpdateApprovalStatus: %v", err)
	}
	story, _ := store.GetByName(epicID, "my-story")
	if story.ApprovalStatus != domain.ApprovalPending {
		t.Errorf("ApprovalStatus = %q, want %q", story.ApprovalStatus, domain.ApprovalPending)
	}
}

File: apps/orchestrator/src/db/stores/__tests__/story-store.test.ts
import { describe, test, expect, beforeEach, afterEach } from "bun:test";
import type { Database } from "bun:sqlite";
import { openMemory } from "../../index.ts";
import { EpicStore } from "../epic-store.ts";
import { StoryStore } from "../story-store.ts";
import { DomainError } from "../../../domain/index.ts";

describe("StoryStore", () => {
  let db: Database;
  let epicStore: EpicStore;
  let store: StoryStore;
  let epicId: number;

  beforeEach(() => {
    db = openMemory();
    epicStore = new EpicStore(db);
    store = new StoryStore(db);
    // Create a parent epic for all story tests
    const epic = epicStore.create("test-epic", "parent epic");
    epicId = epic.id;
  });

  afterEach(() => {
    db.close();
  });

  // --- Create ---

  test("Create inserts a new story and returns it", () => {
    const story = store.create(epicId, "my-story", "A test story");

    expect(story.name).toBe("my-story");
    expect(story.description).toBe("A test story");
    expect(story.epic_id).toBe(epicId);
    expect(story.approval_status).toBe("none");
    expect(story.id).toBeGreaterThan(0);
    expect(story.created_at).toBeTruthy();
    expect(story.updated_at).toBeTruthy();
  });

  test("Create throws ErrAlreadyExists on duplicate (epicId, name)", () => {
    store.create(epicId, "dup-story", "first");

    expect(() => store.create(epicId, "dup-story", "second")).toThrow(DomainError);
    try {
      store.create(epicId, "dup-story", "second");
    } catch (err) {
      expect(err).toBeInstanceOf(DomainError);
      expect((err as DomainError).code).toBe("already_exists");
    }
  });

  test("Create allows same name under different epics", () => {
    const epic2 = epicStore.create("other-epic", "another");
    store.create(epicId, "shared-name", "desc1");
    const story2 = store.create(epic2.id, "shared-name", "desc2");

    expect(story2.name).toBe("shared-name");
    expect(story2.epic_id).toBe(epic2.id);
  });

  // --- GetByName ---

  test("GetByName returns the story", () => {
    store.create(epicId, "find-me", "desc");
    const story = store.getByName(epicId, "find-me");

    expect(story.name).toBe("find-me");
    expect(story.description).toBe("desc");
    expect(story.epic_id).toBe(epicId);
  });

  test("GetByName throws ErrNotFound for missing story", () => {
    expect(() => store.getByName(epicId, "nope")).toThrow(DomainError);
    try {
      store.getByName(epicId, "nope");
    } catch (err) {
      expect(err).toBeInstanceOf(DomainError);
      expect((err as DomainError).code).toBe("not_found");
    }
  });

  // --- List ---

  test("List returns all stories for an epic ordered by creation time", () => {
    store.create(epicId, "alpha", "a");
    store.create(epicId, "beta", "b");
    store.create(epicId, "gamma", "c");

    const stories = store.list(epicId);
    expect(stories).toHaveLength(3);
    expect(stories.map((s) => s.name)).toEqual(["alpha", "beta", "gamma"]);
  });

  test("List returns empty array when no stories exist for epic", () => {
    const stories = store.list(epicId);
    expect(stories).toEqual([]);
  });

  test("List only returns stories for the given epic", () => {
    const epic2 = epicStore.create("other-epic", "other");
    store.create(epicId, "story-a", "a");
    store.create(epic2.id, "story-b", "b");

    const stories = store.list(epicId);
    expect(stories).toHaveLength(1);
    expect(stories[0].name).toBe("story-a");
  });

  // --- Update ---

  test("Update changes description and returns updated story", () => {
    store.create(epicId, "upd-story", "old desc");
    const updated = store.update(epicId, "upd-story", "new desc");

    expect(updated.name).toBe("upd-story");
    expect(updated.description).toBe("new desc");
  });

  test("Update throws ErrNotFound for missing story", () => {
    expect(() => store.update(epicId, "ghost", "desc")).toThrow(DomainError);
    try {
      store.update(epicId, "ghost", "desc");
    } catch (err) {
      expect((err as DomainError).code).toBe("not_found");
    }
  });

  // --- Delete ---

  test("Delete removes a story", () => {
    store.create(epicId, "del-story", "bye");
    store.delete(epicId, "del-story");

    expect(() => store.getByName(epicId, "del-story")).toThrow(DomainError);
  });

  test("Delete throws ErrNotFound for missing story", () => {
    expect(() => store.delete(epicId, "ghost")).toThrow(DomainError);
    try {
      store.delete(epicId, "ghost");
    } catch (err) {
      expect((err as DomainError).code).toBe("not_found");
    }
  });

  // --- UpdateApprovalStatus ---

  test("UpdateApprovalStatus sets the status", () => {
    store.create(epicId, "approval-story", "desc");
    store.updateApprovalStatus(epicId, "approval-story", "approved");

    const story = store.getByName(epicId, "approval-story");
    expect(story.approval_status).toBe("approved");
  });

  test("UpdateApprovalStatus throws ErrNotFound for missing story", () => {
    expect(() => store.updateApprovalStatus(epicId, "ghost", "approved")).toThrow(DomainError);
    try {
      store.updateApprovalStatus(epicId, "ghost", "approved");
    } catch (err) {
      expect((err as DomainError).code).toBe("not_found");
    }
  });

  // --- ListSummaries ---

  test("ListSummaries returns empty array when no stories", () => {
    const summaries = store.listSummaries(epicId);
    expect(summaries).toEqual([]);
  });

  test("ListSummaries returns story with default task counts", () => {
    store.create(epicId, "summary-story", "desc");
    const summaries = store.listSummaries(epicId);

    expect(summaries).toHaveLength(1);
    expect(summaries[0].name).toBe("summary-story");
    expect(summaries[0].task_count).toBe(0);
    expect(summaries[0].completed_task_count).toBe(0);
    expect(summaries[0].failed_task_count).toBe(0);
  });

  test("ListSummaries aggregates task counts", () => {
    store.create(epicId, "counted-story", "desc");
    const story = store.getByName(epicId, "counted-story");

    db.run("INSERT INTO tasks (story_id, name, status) VALUES (?, 'task-1', 'completed')", [story.id]);
    db.run("INSERT INTO tasks (story_id, name, status) VALUES (?, 'task-2', 'failed')", [story.id]);
    db.run("INSERT INTO tasks (story_id, name, status) VALUES (?, 'task-3', 'pending')", [story.id]);

    const summaries = store.listSummaries(epicId);
    expect(summaries).toHaveLength(1);
    expect(summaries[0].task_count).toBe(3);
    expect(summaries[0].completed_task_count).toBe(1);
    expect(summaries[0].failed_task_count).toBe(1);
  });

  test("ListSummaries only returns stories for the given epic", () => {
    const epic2 = epicStore.create("other-epic", "other");
    store.create(epicId, "story-a", "a");
    store.create(epic2.id, "story-b", "b");

    const summaries = store.listSummaries(epicId);
    expect(summaries).toHaveLength(1);
    expect(summaries[0].name).toBe("story-a");
  });
});
